### ISSUE TITLE
Password Restrictions

### ISSUE THREAD

**Issue by davoodharun** [AppliedIS/azure-blueprint#7]
- [x] Strong initial password
- [x] Minimum lifetime (1 day)
- [x] Maximum lifetime (60 days)
- [x] Complexity (14 char. length, at least one of each: upper case, lower case, number, special char.)
- [ ] Change entropy (at least 50%)
- [x] Reuse restrictions (24 generations)
- [x] Change at first logon requirement [do not enable in ref. arch.]
- [x] Password strength enforcement
- [x] Storage/transmission encryption

**Comment by davoodharun** [comment]
@bhavyasinghal I have satisfied this with using a .vhd from a preconfigured vm. Waiting on resolution of #42

**Comment by bhavyasinghal** [comment]
@davoodharun I am getting the following error even when I run the PS script in a VM with DC and web services running, and in all our deployment VMs:
Unable to contact the server. This may be because this server does not exist, it is currently down, or it does not have the Active Directory Web Services running.
I am using this commandlet: "Set-ADDefaultDomainPasswordPolicy"
Can you let me know what are you using?
It is only erroring out on the given cmdlet when I am using credentials and Basic auth type which is very weird, given the error.

Can you let me know what you are using in your script?

**Comment by bhavyasinghal** [comment]
@davoodharun : What does this requirement mean when it says do not enable in reference architecture?-----Change at first logon requirement [do not enable in ref. arch.]

**Comment by bhavyasinghal** [comment]
The script satisfies the following restrictions:
 Strong initial password
 Minimum lifetime (1 day)
 Maximum lifetime (60 days)
 Complexity (14 char. length, at least one of each: upper case, lower case, number, special char.)
 Reuse restrictions (24 generations)
 Change at first logon requirement [do not enable in ref. arch.]
 Password strength enforcement
 Storage/transmission encryption

I had raised a question about the following requirement. Please clarify what this "change entropy" means?
Change entropy (at least 50%)

**Comment by bhavyasinghal** [comment]
@davoodharun I will make the changes discussed in the call (for change entropy and Change at first logon) and send the updated script to you to include in your code.

**Comment by jomolesk** [comment]
Related control(s): IA-5.b, IA-5.e, IA-5 (1).a, IA-5 (1).b, IA-5 (1).c, IA-5 (1).d, IA-5 (1).e, IA-5 (1).f, IA-5 (4), IA-5 (7)

IA-5.b: The organization manages information system authenticators by establishing initial authenticator content for authenticators defined by the organization (i.e., initial passwords must meet all complexity requirements). [_All initial authenticator content for accounts created by this Azure Blueprint Solution meet the requirements stated in IA-5 (1)._]

IA-5.e: The organization manages information system authenticators by changing default content of authenticators prior to information system installation. [_All authenticators for components of this Azure Blueprint Solution have been changed from their defaults. Authenticators are customer-specified during deployment of this solution._]

IA-5 (1).a: The information system, for password-based authentication, enforces minimum password complexity of [Assignment: **case sensitive, minimum of 14 characters, and at least one 1 each of upper-case letters, lower-case letters, numbers, and special characters**]. [_A group policy object (GPO) is applied to all Windows virtual machines deployed by this Azure Blueprint Solution. The group policy object is customized to enforce password complexity requirements for local accounts and AD accounts._]

IA-5 (1).b: The information system, for password-based authentication, enforces at least the following number of changed characters when new passwords are created: [Assignment: **at least 50%**]. [_A group policy object (GPO) is applied to all Windows virtual machines deployed by this Azure Blueprint Solution. The group policy object is customized to enforce password change entropy requirements for local accounts and AD accounts._]

IA-5 (1).c: The information system, for password-based authentication, stores and transmits only cryptographically-protected passwords. [_Azure Directory is used to ensure that all passwords are cryptographically protected while stored and transmitted. Passwords stored by Active Directory and locally on deployed Windows virtual machines are automatically hashed as part of built-in security functionality._ **NOTE: Need to confirm that all authentication sessions (e.g., RDP) protect authenticator content when transmitted.**]

IA-5 (1).d: The information system, for password-based authentication, enforces password minimum and maximum lifetime restrictions of [Assignment: **one day minimum, sixty day maximum**]. [_A group policy object (GPO) is applied to all Windows virtual machines deployed by this Azure Blueprint Solution. The group policy object is customized to enforce restrictions on passwords that enforce minimum (1 day) and maximum (60 days) lifetime restrictions for local accounts and AD accounts._]

IA-5 (1).e: The information system, for password-based authentication, prohibits password reuse for [Assignment: **24**] generations. [_A group policy object (GPO) is applied to all Windows virtual machines deployed by this Azure Blueprint Solution. The group policy object is customized to enforce  restrictions on reuse conditions (24 passwords) for local accounts and AD accounts._]

IA-5 (1).f: The information system, for password-based authentication, allows the use of a temporary password for system logons with an immediate change to a permanent password. [_**NOTE: need to add implementation details if we deploy any temporary passwords (we should try to avoid doing this).**_]

IA-5 (4): The organization employs automated tools to determine if password authenticators are sufficiently strong to satisfy [Assignment: organization-defined requirements]. This control enhancement focuses on the creation of strong passwords and the characteristics of such passwords (e.g., complexity) prior to use. [_User accounts deployed with this Azure Blueprint Solution include AD and local user accounts. Both of these provide mechanisms that force compliance with established password requirements in order to create an initial password and during password changes._]

IA-5 (7): The organization ensures that unencrypted static authenticators are not embedded in applications or access scripts or stored on function keys. [_There is no use of unencrypted static authenticators embedded in applications, access scripts, or function keys deployed by this Azure Blueprint Solution. Any script or application that uses an authenticator makes a call to an Azure Key Vault container prior to each use. Access to Azure Key Vault containers is audited, which allows detection of violations of this prohibition if a service account is used to access a system without a corresponding call to the Azure Key Vault container._] (Also see issue #9.)

**Comment by jomolesk** [comment]
@davoodharun This needs only small changes to verify the initial passwords meet the complexity requirements: https://gallery.technet.microsoft.com/scriptcenter/Verify-password-complexity-c9e6f42f

**Comment by bhavyasinghal** [comment]
@davoodharun @jmmcnj I uploaded the current script in Blocks folder as discussed. There is still unresolved issue with this script as I had emailed.

**Comment by davoodharun** [comment]
closing this issue with #53.

Creating a new issue for the change entropy requirement.

**Comment by bhavyasinghal** [comment]
@reetikatech 
Please test with following instructions:

Step 1:

1) Login to domain controller VM

2) From Start Menu- go to active directory users and computers.

3) Select domain->users folder

4) On the right pane, right click->new->user.

5) Create a new user by providing details such as first name, last name, userlogon name.

6) Click Next-> create a new password and confirm the password. Uncheck "user must change password at next logon".

7) The password you gave must: "have 14 char length, and AT LEAST 3 out of the 4 type of given characters: upper case, lower case, number, special char".

8) Click Next-> click finish.



Step 2:
1) Login to domain controller VM

2) From Start Menu->Run->GPEDIT.msc->ok

3)Local computer policy->computer configuration->windows settings->security settings->account policies->password policy.

4) The following should be the values:

- Enforce password history: 24 passwords remembered
- Maximum password age: 60 days
- Minimum password age: 1 day
- Minimum password length: 14 characters
- Password must meet complexity requirement: Enabled
- Store Passwords using reversible encryption: Enabled

**Comment by reetikatech** [comment]
tested and verified by Reetika by executing the script and using the steps above as well.
change entropy is still to be work on

**Comment by davoodharun** [comment]
Need to confirm that all authentication sessions (e.g., RDP) protect authenticator content when transmitted.

**Comment by davoodharun** [comment]
# Password Restrictions
Password Restriction controls are met by applying a domain Group Policy that is applied to every computer in the domain. The domain Group Policy is applied by a PowerShell script deployed through a Windows Azure Custom Script Extension. The script can be found [here](../../postdeploy/setPasswordPolicy.ps1). This provides a set of restrictions for any profiles created after deployment on any machine.

Validation is done for initial administrator account passwords through a PowerShell script that can be found [here](../../predeploy/checkPassword.ps1).

## Configuration
#### Group Domain Policy
- Configurations to the Group Domain Policy must be made in [setPasswordPolicy.ps1](../../postdeploy/setPasswordPolicy.ps1)).

#### Initial Admin Account
- Configurations to initial admin account password restrictions can be made in [checkPassword.ps1](../../predeploy/checkPassword.ps1).

## Group Domain Policy
Restrictions applied:
- Minimum lifetime (1 day)
- Maximum lifetime (60 days)
- Complexity (14 char. length, at least one of each: upper case, lower case, number, special char.)
- Reuse restrictions (24 generations)
- Storage/transmission encryption
- Strong initial password
- Password strength enforcement
- [RDP connection encryption (high)](https://technet.microsoft.com/en-us/library/ff458357.aspx)

\* Password change entropy is not applied
  
## Initial admin user accounts.
- Complexity (14 char. length, at least one of each: upper case, lower case, number, special char.)
- No blank passwords
- No spaces
